File: main-app/backend/app/models/brain.py
"""
Brain model — an autonomous AI employee.

Each Brain is configured from a template, goes through a setup interview,
connects to external accounts, and then executes tasks autonomously.
"""

from datetime import datetime
from typing import Optional
from uuid import uuid4

from sqlalchemy import String, DateTime, Text, JSON, ForeignKey, Boolean, Integer
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.database import Base


class Brain(Base):
    """An autonomous AI employee (Brain)."""

    __tablename__ = "brains"

    brain_id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid4())
    )
    user_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.user_id", ondelete="CASCADE"),
        nullable=False, index=True,
    )
    tenant_id: Mapped[Optional[str]] = mapped_column(
        String(36), ForeignKey("tenants.tenant_id"), nullable=True,
    )
    template_id: Mapped[Optional[str]] = mapped_column(
        String(36), ForeignKey("brain_templates.template_id"), nullable=True,
    )

    # Identity
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    brain_type: Mapped[str] = mapped_column(
        String(50), nullable=False, default="custom"
    )  # job_searcher | trader | social_media | coder | personal_finance | custom
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    icon: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    color: Mapped[Optional[str]] = mapped_column(String(7), nullable=True)

    # Setup interview state
    setup_status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="pending"
    )  # pending | interview | ready | error
    setup_step: Mapped[int] = mapped_column(Integer, default=0)
    setup_answers: Mapped[Optional[dict]] = mapped_column(JSON, nullable=True)

    # System prompt built from template + setup answers
    system_prompt: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    # Autonomy level
    autonomy_level: Mapped[str] = mapped_column(
        String(20), nullable=False, default="supervised"
    )  # supervised | semi_auto | full_auto

    # Runtime state
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="inactive"
    )  # inactive | active | paused | error
    is_active: Mapped[bool] = mapped_column(Boolean, default=False)
    last_active_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)

    # Daily limits
    max_daily_tasks: Mapped[int] = mapped_column(Integer, default=50)
    max_daily_cost_cents: Mapped[int] = mapped_column(Integer, default=500)
    tasks_today: Mapped[int] = mapped_column(Integer, default=0)
    cost_today_cents: Mapped[int] = mapped_column(Integer, default=0)

    # Brain-type-specific configuration
    config_json: Mapped[Optional[dict]] = mapped_column(JSON, default=dict)
    metadata_json: Mapped[Optional[dict]] = mapped_column(JSON, default=dict)

    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow
    )

    # Relationships
    connected_accounts = relationship(
        "ConnectedAccount", back_populates="brain", cascade="all, delete-orphan"
    )
    tasks = relationship(
        "BrainTask", back_populates="brain", cascade="all, delete-orphan"
    )
    schedules = relationship(
        "BrainSchedule", back_populates="brain", cascade="all, delete-orphan"
    )
    monitors = relationship(
        "BrainMonitor", back_populates="brain", cascade="all, delete-orphan"
    )
    activities = relationship(
        "BrainActivity", back_populates="brain", cascade="all, delete-orphan"
    )
    pipeline_items = relationship(
        "PipelineItem", back_populates="brain", cascade="all, delete-orphan"
    )

    def to_dict(self) -> dict:
        return {
            "brain_id": self.brain_id,
            "user_id": self.user_id,
            "tenant_id": self.tenant_id,
            "template_id": self.template_id,
            "name": self.name,
            "brain_type": self.brain_type,
            "description": self.description,
            "icon": self.icon,
            "color": self.color,
            "setup_status": self.setup_status,
            "setup_step": self.setup_step,
            "setup_answers": self.setup_answers,
            "autonomy_level": self.autonomy_level,
            "status": self.status,
            "is_active": self.is_active,
            "last_active_at": self.last_active_at.isoformat() if self.last_active_at else None,
            "max_daily_tasks": self.max_daily_tasks,
            "max_daily_cost_cents": self.max_daily_cost_cents,
            "tasks_today": self.tasks_today,
            "cost_today_cents": self.cost_today_cents,
            "config": self.config_json or {},
            "metadata": self.metadata_json or {},
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

File: main-app/frontend/src/types/brain.ts
// Brain platform types

export interface BrainTemplate {
  template_id: string
  slug: string
  name: string
  description: string
  icon: string
  color: string
  category: string
  interview_questions: InterviewQuestion[]
  required_accounts: string[]
  optional_accounts: string[]
  available_tools: string[]
  is_builtin: boolean
  created_at: string | null
}

export interface InterviewQuestion {
  key: string
  question: string
  type: 'text' | 'textarea' | 'select' | 'multi_select' | 'multiselect' | 'multi_text' | 'number' | 'boolean' | 'connect_account' | 'file_upload'
  required?: boolean
  options?: (string | { label: string; value: unknown })[]
  placeholder?: string
  provider?: string  // for connect_account type
  accept?: string    // for file_upload type
  condition?: { key: string; values?: string[]; value?: string }
}

export interface Brain {
  brain_id: string
  user_id: string
  tenant_id: string | null
  template_id: string | null
  name: string
  brain_type: string
  description: string | null
  icon: string | null
  color: string | null
  setup_status: 'pending' | 'interview' | 'ready' | 'error'
  setup_step: number
  autonomy_level: 'supervised' | 'semi_auto' | 'full_auto'
  status: 'inactive' | 'active' | 'paused' | 'error'
  is_active: boolean
  max_daily_tasks: number
  max_daily_cost_cents: number
  tasks_today: number
  cost_today_cents: number
  config: Record<string, unknown>
  created_at: string | null
  updated_at: string | null
}

export interface InterviewState {
  brain_id: string
  setup_status: string
  current_step: number
  total_steps: number
  current_question: InterviewQuestion | null
  answers: Record<string, unknown>
  is_complete: boolean
}

export interface ConnectedAccount {
  account_id: string
  brain_id: string
  provider: string
  provider_display_name: string
  account_identifier: string | null
  auth_type: string
  status: string
  status_message: string | null
  is_active: boolean
  last_used_at: string | null
  last_verified_at: string | null
  created_at: string | null
}

export interface BrainTask {
  task_id: string
  brain_id: string
  schedule_id: string | null
  pipeline_item_id: string | null
  parent_task_id: string | null
  task_type: string
  title: string
  instructions: string | null
  status: 'pending' | 'awaiting_approval' | 'running' | 'completed' | 'failed' | 'cancelled'
  priority: number
  trigger: string
  requires_approval: boolean
  result_summary: string | null
  error: string | null
  cost_cents: number
  queued_at: string | null
  started_at: string | null
  completed_at: string | null
}

export interface ApprovalRequest {
  request_id: string
  brain_id: string
  task_id: string | null
  action_type: string
  action_summary: string
  action_data: Record<string, unknown> | null
  status: 'pending' | 'approved' | 'denied' | 'expired' | 'auto_approved'
  expires_at: string | null
  resolved_at: string | null
  denial_reason: string | null
  created_at: string | null
}

export interface BrainMonitor {
  monitor_id: string
  brain_id: string
  name: string
  monitor_type: string
  target_url: string | null
  target_config: Record<string, unknown>
  check_interval_minutes: number
  trigger_condition: string
  trigger_action: string
  notification_channels: string[]
  is_active: boolean
  last_check_at: string | null
  trigger_count: number
  created_at: string | null
}

export interface PipelineItem {
  item_id: string
  brain_id: string
  pipeline_type: string
  title: string
  external_url: string | null
  stage: string
  stage_order: number
  data_json: Record<string, unknown>
  history_json: { stage: string; timestamp: string }[]
  is_starred: boolean
  is_archived: boolean
  created_at: string | null
  updated_at: string | null
}

export interface BrainActivity {
  activity_id: string
  brain_id: string
  task_id: string | null
  activity_type: string
  title: string
  description: string | null
  detail_json: Record<string, unknown> | null
  severity: 'info' | 'success' | 'warning' | 'error'
  created_at: string | null
}

export interface DashboardOverview {
  brains: {
    total: number
    active: number
    paused: number
    inactive: number
  }
  today: {
    tasks_completed: number
    cost_cents: number
  }
  pending_approvals: number
}

export interface BrainStats {
  brain: Brain
  tasks: { total: number; by_status: Record<string, number> }
  pipeline: { total: number; by_stage: Record<string, number> }
  connected_accounts: number
  active_monitors: number
  total_activities: number
}

// Icon mapping for brain types
export const BRAIN_ICONS: Record<string, string> = {
  job_searcher: '💼',
  trader: '📈',
  social_media: '📱',
  coder: '💻',
  personal_finance: '💰',
  custom: '🧠',
}

export const BRAIN_STATUS_COLORS: Record<string, string> = {
  active: 'text-green-500',
  paused: 'text-yellow-500',
  inactive: 'text-gray-400',
  error: 'text-red-500',
}

export const TASK_STATUS_COLORS: Record<string, string> = {
  pending: 'bg-gray-100 text-gray-600',
  awaiting_approval: 'bg-yellow-100 text-yellow-700',
  running: 'bg-blue-100 text-blue-700',
  completed: 'bg-green-100 text-green-700',
  failed: 'bg-red-100 text-red-700',
  cancelled: 'bg-gray-100 text-gray-500',
}
